data_node/upload_manager.go

```diff
@@ -102,8 +106,86 @@ func (um *UploadManager) handleInitialUpload(w http.ResponseWriter, r *http.Requ
 		return
 	}
 
-	w.WriteHeader(http.StatusCreated)
 	w.Header().Set("ID", id)
+	w.Header().Set("Max-Request-Size", strconv.FormatInt(um.maxChunkSize, 10))
+	w.WriteHeader(http.StatusCreated)
+}
+
+// handleAppendUpload is a function responsible for handling the first upload request
+func (um *UploadManager) handleAppendUpload(w http.ResponseWriter, r *http.Request) {
+	r.Body = http.MaxBytesReader(w, r.Body, um.maxChunkSize)
+	log.Println(um.logPrefix, r.RemoteAddr, "Received APPEND request")
+
+	expectedHeaders := []string{"Content-Length", "Offset", "ID"}
+	err := um.validateUploadHeaders(&r.Header, expectedHeaders...)
+
+	if err != nil {
+		log.Println(um.logPrefix, r.RemoteAddr, err)
+		handleRequestError(w, http.StatusBadRequest, err.Error())
+		return
+	}
+
+	id := r.Header.Get("ID")
+	if !um.validateIDExistance(id) {
+		log.Println(um.logPrefix, r.RemoteAddr, "ID not found")
+		handleRequestError(w, http.StatusNotFound, "ID not found")
+		return
+	}
+
+	chunkSize, err := strconv.ParseInt(r.Header.Get("Content-Length"), 10, 64)
+	if errors.IsError(err) || chunkSize <= 0 {
+		log.Println(um.logPrefix, r.RemoteAddr, "Invalid chunk size", r.Header.Get("Content-Length"))
+		handleRequestError(w, http.StatusBadRequest, "Invalid chunk size")
+		return
+	}
+
+	offset, err := strconv.ParseInt(r.Header.Get("Offset"), 10, 64)
+	if errors.IsError(err) || !um.validateFileOffset(id, offset, chunkSize) {
+		log.Println(um.logPrefix, r.RemoteAddr, "Invalid file offset", r.Header.Get("Offset"))
+		w.Header().Set("Offset", strconv.FormatInt(um.fileBase[id].Offset, 10))
+		handleRequestError(w, http.StatusBadRequest, "Invalid offset")
+		return
+	}
+
+	file := um.fileBase[id]
+	filePath := path.Join(file.Path, file.Name)
+
+	lock := fslock.New(filePath)
+	lock.Lock()
+	defer lock.Unlock()
```

Comment:
It's better to move the unlock after the code that needs the lock since there is other code after this that needs a different lock. This could cause a deadlock and it holds the lock for longer than needed.

Reply:
@ahmibr What about this?